Improve config (closes #810)

ire-and-curses opened the pull request:
This closes #810 by cleaning up the redundancy in the Horizon command line option handling.

A new array, `configOpts`, is used to fully define the command line parameters. To add a new command line parameter, it should be sufficient to add a line to `configOpts`, and a new field to the `horizon.Config` struct.

Optionally, the developer can hook a custom function to the new `configOpts` line for validation and/or transformation of the incoming parameter.

I wanted to avoid invasive changes, but during this refactoring I found a few inconsistencies which are worth pointing out:
1) Unused option `ruby-horizon-url` (I removed)
2) Incomplete configuration for `skip-cursor-update` (I added missing config)
3) Redundant viper defaults for `port` and `history-retention-count` (I removed)
4) `friendbot-url` is stored as a `url.URL`, but other URLs (e.g. `db-url`, `stellar-core-db-url`, `stellar-core-url`, `redis-url`, `sentry-dsn` etc.) are just strings. Should we change this in the future?
5) `tls-cert` and `tls-key` have custom checking to ensure both are provided, but there are other pairs that could be similarly checked, e.g. `redis-limit-redis-key` and `redis-url`, `loggly-token` and `loggly-tag` (easy to do).
6) Inconsistent names could be cleaned up, e.g. there is no reason to map `history-stale-threshold` to `stale-threshold`, or `db-url` to the rather generic env variable `DATABASE_URL`.
7) Some numbers are `uint` but others are not. Not sure what the logic is for the `uints`.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -260,77 +191,99 @@ func initConfig() {
-	ll, err := logrus.ParseLevel(viper.GetString("log-level"))
+// configOpts defines the complete flag configuration for horizon. Add a new line here to connect a new field in the horizon.Config struct
+var configOpts = []configOption{
+	configOption{name: "db-url", envVar: "DATABASE_URL", configKey: &c.DatabaseURL, flagDefault: "", required: true, usage: "horizon postgres database to connect with"},
```
Ok, will do :) I went the other way because I preferred the denser representation (easier to compare columns). This was also the reason I did `c = config` - to save horizontal space. I'm happy to change it though.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -14,238 +14,169 @@ import (
 
-	viper.BindPFlags(rootCmd.PersistentFlags())
+// setValue sets a value in the global config, using a custom function, if one was provided.
+func (co *configOption) setValue() error {
```
good point, removed.

bartekn commented:
PS. you can consider closing https://github.com/stellar/go/issues/817 in this PR.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -260,77 +191,99 @@ func initConfig() {
 	}
 
-	log.DefaultLogger.Level = ll
+	for i := range configOpts {
```
ok, I changed `configOpts` to type `[]*configOption` to allow this.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -260,77 +191,99 @@ func initConfig() {
+		// Unless overriden, default to a transform like tls-key -> TLS_KEY
+		if co.envVar == "" {
+			co.envVar = strutils.KebabToConstantCase(co.name)
+			viper.BindEnv(co.name, co.envVar)
```
good catch

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -14,238 +14,169 @@ import (
+	name           string              // e.g. "db-url"
+	envVar         string              // e.g. "DATABASE_URL". Defaults to uppercase/underscore representation of name
+	flagType       flagType            // e.g. boolFlag
+	flagDefault    interface{}         // A default if no option is provided. Set to "" if no default
```
I use "" because `flagDefault` is used to infer the type. We never set empty defaults for other types, so we can use the empty string as a type indicator. This allows us to determine the option type without the need to explicitly set it, saving a lot of extra configuration.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -14,238 +14,169 @@ import (
+// setSimpleValue sets the value of a configOption's configKey, based on the configOption's default type.
+func (co *configOption) setSimpleValue() {
+	if co.configKey != nil {
+		switch co.flagDefault.(type) {
```
No, this was intended. The idea is to use the default to infer the type. That's why the type is never set in any of the `configOption`s. In fact, `flagType` should be removed from the `configOption` struct.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -260,77 +191,99 @@ func initConfig() {
+			viper.BindEnv(co.name, co.envVar)
 		}
+		// Initialise the persistent flags
+		co.setFlag()
```
I like this idea. But I think we can't move the contents of both loops (see next comment) to the same method. They are executed at different times. This `for` loop only happens before the app is served.

ire-and-curses replied:
I moved this loop body to `func (co *configOption) init()`.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -260,77 +191,99 @@ func initConfig() {
+	// Check all required args were provided - needed for migrations check
+	for i := range configOpts {
+		co := &configOpts[i]
+		co.require()
```
I like this idea. But I think we can't move the contents of both loops to the same method. They are executed at different times. This `require` step only happens when the app is executed.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -260,77 +191,99 @@ func initConfig() {
+	// Validate options that should be provided together
+	validateTLS(tlsProvided)
+
+	config = c
```
This was just an alias to keep the `configOpts` "table" shorter. I will remove it and write out the rows as long-form entries.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -14,238 +14,169 @@ import (
+	if err != nil {
+		stdLog.Fatalf("Could not parse log-level: %v", viper.GetString(co.name))
+	}
+	log.DefaultLogger.Level = ll
```
Ok. I was trying to keep all the config work in one place for convenience, but I agree with your point. For these special options, there is additional configuration that should be  in the `initConfig()`, since they change the state of something other than our `config` object.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -260,77 +191,99 @@ func initConfig() {
+		co.setValue()
 	}
+	// Validate options that should be provided together
+	validateTLS(tlsProvided)
```
My aim was to remove all calls to literal strings, e.g. replace things like `viper.getString("log-file")` with `viper.getString(co.name)`. Now only the migration checks still look up the parameters by literal strings. I can generalise this though to be a function that takes two config objects, and confirms both are present, which would be simpler, and would also let me check other pairs.

ire-and-curses commented on services/horizon/main.go:
```diff
@@ -14,238 +14,169 @@ import (
-		"the maximum number of assets on the path in `/paths` endpoint",
-	)
+// configOption is a complete description of the configuration of a command line option
+type configOption struct {
```
Ok, moved to `/support/config/`.

ire-and-curses commented:
Can't change `MaxDBConnections` to `uint`, code expects `int`.
Most other changes made. Please take another look.

bartekn commented on services/horizon/main.go:
```diff
@@ -2,61 +2,222 @@ package main
+		Name:        "db-url",
+		EnvVar:      "DATABASE_URL",
+		ConfigKey:   &config.DatabaseURL,
+		FlagDefault: "",
```
Continuing [this discussion](https://github.com/stellar/go/pull/834#discussion_r251362820) here.
> I use "" because flagDefault is used to infer the type.

I think it shouldn't be used to get type. It's like embedding two values in a single field. Instead we should use [`types.*`](https://golang.org/pkg/go/types/#BasicKind) (`types.Bool`, `types.Uint`, `types.String`, etc). The advantage is that it's clear what type we are dealing with and we can omit `FlagDefault` if it's a _zero_ value.

tomquisel replied:
This suggestion makes sense to me too

ire-and-curses replied:
Done. Unfortunately `pflags` library requires a default value for all flags (see e.g. [pflag.Bool](https://godoc.org/github.com/spf13/pflag#Bool)), so I wrote a helper `setDefault` to add the empty string if a `FlagDefault` is not explicitly declared.

bartekn commented on services/horizon/main.go:
```diff
@@ -230,107 +250,32 @@ func initApp(cmd *cobra.Command, args []string) *horizon.App {
-		stdLog.Fatal("Invalid config: stellar-core-url is blank.  Please specify --stellar-core-url on the command line or set the STELLAR_CORE_URL environment variable.")
+	// Initialise and validate the global configuration
+	for _, co := range configOpts {
+		co.SetValue()
```
Continuing [this discussion](https://github.com/stellar/go/pull/834#discussion_r251367094) here.

> But I think we can't move the contents of both loops to the same method. They are executed at different times. This require step only happens when the app is executed.

Any reason why `Require` and `SetValue` can't be called along with other init steps in one place? You can actually move:
```go
	for _, co := range configOpts {
		co.Init(rootCmd)
	}
```
to `initConfig()`.

ire-and-curses replied:
These can't be combined - they are needed at different stages of the option parsing lifecycle. I think this is why the code was originally written with both `init()` and `initConfig()` methods. You can test for yourself - if you move the code into a single loop, it will fail and display help text.

bartekn replied:
So I pulled the branch locally and indeed it wasn't possible with `viper.BindPFlags` as it required all flags to be set upfront. However, after switching to `viper.BindPFlag` (note: no `s`) it was possible. I actually pushed this commit: https://github.com/stellar/go/pull/834/commits/75b8658fd5662c321b7a1490f188edc17a287083 hope you don't mine. I also removed `flagType` as it was used in only a single method I actually changed and added error handling for `viper` methods that return errors.

ire-and-curses replied:
Looks great, thanks for those improvements! :+1:  Are we good to merge now?

bartekn commented on services/horizon/main.go:
```diff
@@ -230,107 +250,32 @@ func initApp(cmd *cobra.Command, args []string) *horizon.App {
 	}
+	// Validate options that should be provided together
+	validateBothOrNeither("tls-cert", "tls-key")
+	validateBothOrNeither("rate-limit-redis-key", "redis-url")
```
This will ignore values set via environment variables. Values are in `Config` object at this point so let's use `Config` values.

ire-and-curses replied:
good point, done.

bartekn commented on support/config/config_option.go:
```diff
@@ -0,0 +1,142 @@
+		stdLog.Fatalf("Could not parse log-level: %v", viper.GetString(co.Name))
+	}
+	*(co.ConfigKey.(*logrus.Level)) = ll
+}
```
I think this package shouldn't have dependencies other than standard packages, stellar/go and viper. Consider that we are building a new app that's using `ConfigOption` but doesn't use rate limiting. It will have to download `github.com/throttled/throttled` anyway.

I think that `SetLogLevel` and `SetRateLimit` should be moved to `horizon.main.go` - can be used as inline functions, ex.:
```go
	&support.ConfigOption{
		Name:           "log-level",
		ConfigKey:      &config.LogLevel,
		FlagDefault:    "info",
		CustomSetValue: func(co *ConfigOption) {
			ll, err := logrus.ParseLevel(viper.GetString(co.Name))
			if err != nil {
				stdLog.Fatalf("Could not parse log-level: %v", viper.GetString(co.Name))
			}
			*(co.ConfigKey.(*logrus.Level)) = ll
		},
		Usage:          "minimum log severity (debug, info, warn, error) to log",
	},
```

ire-and-curses replied:
ok, I moved this. I think the resulting inline functions are pretty ugly, but they do keep the config in one place, so I'm ok with it.

bartekn commented on services/horizon/main.go:
```diff
@@ -230,107 +277,27 @@ func initApp(cmd *cobra.Command, args []string) *horizon.App {
-		os.Exit(1)
-	}
+	// Validate options that should be provided together
+	validateBothOrNeither(config.TLSCert, config.TLSKey)
```
Looks like this function sends values of fields that are then used as parameter names. Other than that: when I was suggesting this [here](https://github.com/stellar/go/pull/834#discussion_r251795392) I was wrong about two things: 1) I didn't consider that the types can be different that makes this more difficult but also 2) I suggested that environment variables were ignores - this is probably not the case. Could you just test the previous version of this function works with env variables? If not, it's 👍 to me.

ire-and-curses replied:
Good catch! Confirmed that switching back to strings like `"tls-cert"` fixes the issue, and verified that setting env var like `TLS_CERT` works too.

bartekn commented on support/config/config_option.go:
```diff
@@ -0,0 +1,130 @@
+		return err
+	}
+	co.Require()
+	co.SetValue()
```
The methods above are now private so we can change lowercase the first letters. Sorry forgot to do it in my commit.

ire-and-curses replied:
These are now called from main.go again, so left them public.

bartekn commented on support/config/config_option.go:
```diff
@@ -0,0 +1,130 @@
+
+// Require checks that a required string configuration option is not empty, raising a user error if it is.
+func (co *ConfigOption) Require() {
+	if co.Required == true && viper.GetString(co.Name) == "" {
```
Can you confirm if `viper.GetString` works fine if the value is passed via environment variable?

ire-and-curses replied:
Confirmed works by testing with `DATABASE_URL`.

bartekn commented on support/config/config_option.go:
```diff
@@ -0,0 +1,130 @@
+		stdLog.Fatal("Non-string options require a default to be set")
+	}
+	co.FlagDefault = ""
+}
```
Can we move this to `setFlag` and remove this method? If it's under `case types.String:` we can compress it to:
```go
if co.FlagDefault == nil {
    co.FlagDefault = ""
}
```
so 3 lines vs 11 (with comments).

ire-and-curses replied:
great! Done

bartekn commented on support/config/config_option.go:
```diff
@@ -0,0 +1,130 @@
+	Required       bool                // Whether this option must be set for Horizon to run
+	Usage          string              // Help text
+	CustomSetValue func(*ConfigOption) // Optional function for custom validation/transformation
+	ConfigKey      interface{}         // Pointer to the final key in the horizon.Config struct
```
This can be used outside horizon in the future so update the comment.

ire-and-curses commented:
Ok, fixed issue with location of pflags binding, addressed other comments. Please take another look, and I'd appreciate a quick test to make sure this runs for you as well.
